File: insert-delete-getrandom-o1-duplicates-allowed/solution.cpp
class RandomizedCollection {
private:
    vector<int> a;
    unordered_map<int, unordered_set<int>> idx;

public:
    RandomizedCollection() {
        a.clear();
        idx.clear();
    }

    bool insert(int val) {
        idx[val].insert(a.size());
        a.push_back(val);
        return idx[val].size() == 1;
    }

    bool remove(int val) {
         if (idx[val].empty()) {
            return false;
        }
        auto remove_it = idx[val].begin();
        int remove = *remove_it;
        int last = a.back();
        a[remove] = last;
        idx[val].erase(remove_it);
        idx[last].insert(remove);
        idx[last].erase(a.size() - 1);
        a.pop_back();
        return true;
    }

    int getRandom() {
        return a[rand() % a.size()];
    }
};

File: insert-delete-getrandom-o1-duplicates-allowed/solution.py
class RandomizedCollection:

    def __init__(self):
        self.a = []
        self.idx = defaultdict(set)

    def insert(self, val: int) -> bool:
        self.idx[val].add(len(self.a))
        self.a.append(val)
        return len(self.idx[val]) == 1

    def remove(self, val: int) -> bool:
        if len(self.idx[val]) == 0:
            return False
        remove, last = self.idx[val].pop(), self.a[-1]
        self.a[remove] = last
        self.idx[last].add(remove)
        self.a.pop()
        self.idx[last].remove(len(self.a))
        return True

    def getRandom(self) -> int:
        return random.choice(self.a)